src/ocamldep.ml

```diff
@@ -124,17 +140,55 @@ let rules ~(ml_kind:Ml_kind.t) ~dir ~modules
       match Module.file ~dir unit ml_kind with
       | None -> Build.return []
       | Some file ->
-        let ocamldep_output = Path.extend_basename file ~suffix:".d" in
+        let all_deps_path file =
+          Path.extend_basename file ~suffix:".all-deps"
+        in
         let context = SC.context sctx in
+        let all_deps_file = all_deps_path file in
+        let ocamldep_output = Path.extend_basename file ~suffix:".d" in
+        let single_deps_output = Path.extend_basename file ~suffix:".deps" in
+        let read_deps_file path =
+          Build.lines_of path
+          >>^ parse_module_names
+                ~dir ~unit ~modules ~alias_module
+                ~lib_interface_module
+                ~check:false
+        in
         if not (Module.Name.Set.mem already_used unit.name) then
-          SC.add_rule sctx
-            (Build.run ~context (Ok context.ocamldep)
-               [A "-modules"; Ml_kind.flag ml_kind; Dep file]
-               ~stdout_to:ocamldep_output);
-        Build.memoize (Path.to_string ocamldep_output)
-          (Build.lines_of ocamldep_output
-           >>^ parse_deps ~dir ~file ~unit ~modules ~alias_module
-                 ~lib_interface_module))
+          begin
+            SC.add_rule sctx
+              ( Build.run ~context (Ok context.ocamldep)
+                  [A "-modules"; Ml_kind.flag ml_kind; Dep file]
+                  ~stdout_to:ocamldep_output
+              );
+            SC.add_rule sctx
+              ( Build.lines_of ocamldep_output
+                >>^ parse_deps
+                      ~dir ~file ~unit ~modules ~alias_module
+                      ~lib_interface_module
+                >>^ to_deps_format
+                >>> Build.write_file_dyn single_deps_output
+              );
+            let build_paths dependencies =
+              let dependency_file_path m =
+                let path =
+                  match Module.file ~dir m Ml_kind.Intf with
+                  | Some x -> Some x
+                  | None when Option.is_some alias_module -> None
```

Comment:
I had a look and I understand why we need this: `alias_modules` is in `modules` so we setup the rule to call ocamldep on it and given it's generated contents, it always reports dependencies on every other modules. Additionally, we manually add a dependency from every other module to `alias_module`, so we end up with a circular dependency.

We didn't notice the issue before this PR as we manually override the binding for `alias_module` just after (line 193). What we should do instead is add a test at the beginning of this function (line 140) to make sure we don't generate rules for `alias_module` and just return `Build.return []` instead. `dependency_file_path` should return `None` when `m` is `alias_module`.

Reply:
I moved that part to the main `Module.Name.Map.map modules` call as you suggested. There's some duplication between the two `Module.file` calls since they're both coupled to a match on `alias_module`, but that might be early to extract an helper function.